Framework/Algorithms/src/XrayAbsorptionCorrection.cpp

```diff
@@ -0,0 +1,234 @@
+  double radians = M_PI * (degrees / 180);
+  return radians;
+}
+
+/**
+ * Calculates the position of the detector.
+ * @param detectorAngle double of angle of detector
+ * @param detectorDistance distance betweeen sample and detector
+ * @return V3D object of position of detector
+ */
+Kernel::V3D
+XrayAbsorptionCorrection::calculateDetectorPos(double detectorAngle,
+                                               double detectorDistance) {
+  detectorDistance = detectorDistance / 100;
+  double x = detectorDistance / std::tan(degreesToRadians(detectorAngle));
+  if (detectorAngle > 180) {
+    detectorDistance = -detectorDistance;
+  }
+  Kernel::V3D detectorPos = {x, detectorDistance, 0.};
+  return detectorPos;
+}
+
+/**
+ * normalise moun intensity to 1.
+ * @param muonIntensity A MantidVec which contains the intensity of muons as a
+ * function  of depth
+ * @return A vector of doubles which contains the normalised muon intensity
+ */
+std::vector<double> XrayAbsorptionCorrection::normaliseMuonIntensity(
+    MantidVec const &muonIntensity) {
+  double sum_of_elems{0};
+  for (double elem : muonIntensity) {
+    sum_of_elems += elem;
+  }
+
+  std::vector<double> normalisedMuonIntensity;
+  for (double x : muonIntensity) {
+    normalisedMuonIntensity.push_back(x / sum_of_elems);
+  }
+  return normalisedMuonIntensity;
+}
+
+/**
+ * Calculate the muon implantation position in the sample.
+ * @param muonProfile A reference to the muon profile workspace
+ * @param inputWS A reference to the input workspace
+ * @param detectorDistance a double representing the distance from the sample to
+ * detector
+ * @return A vector of V3D objects that represent the position of muons
+ */
+std::vector<Kernel::V3D> XrayAbsorptionCorrection::calculateMuonPos(
+    API::MatrixWorkspace_sptr &muonProfile, API::MatrixWorkspace_sptr inputWS,
+    double detectorDistance) {
+  const MantidVec muonDepth = muonProfile->readX(0);
+  Kernel::V3D muonPoint = {0., detectorDistance, 0.};
+  Kernel::V3D toStart = {0, -1, 0};
+  const Geometry::IObject *shape = &inputWS->sample().getShape();
+  Geometry::Track muonPath = Geometry::Track(muonPoint, toStart);
+  shape->interceptSurface(muonPath);
+  if (muonPath.count() == 0) {
+    throw std::runtime_error("No valid solution, check shape parameters, Muon "
+                             "depth profile and detector distance");
+  }
+  Kernel::V3D intersection = muonPath.cbegin()->entryPoint;
+  double sampleDepth = intersection[1];
+  std::vector<Kernel::V3D> muonPos;
+  for (auto depth : muonDepth) {
+    Kernel::V3D pos = {0., sampleDepth - (depth / 100), 0.};
+    muonPos.push_back(pos);
+  }
+  return muonPos;
+}
+
+/**
+ * Calculate the mid point of a bin in workspace.
+ * @param energyVector a MantidVec which contains edges of bins
+ * @return A vector of doubles of bin mid points
+ */
+std::vector<double> XrayAbsorptionCorrection::calculateMidPointOfBin(
+    MantidVec const &energyVector) {
```

Comment:
is the energy vector always the x values for the histogram bins?

Reply:
yes

Reply:
you can then convert the workspace to point data and pull the x values from that

Reply:
https://docs.mantidproject.org/nightly/algorithms/ConvertToPointData-v1.html and https://docs.mantidproject.org/nightly/concepts/HistogramData.html